cobalt/js_profiler/profiler.h

```diff
@@ -0,0 +1,82 @@
+// Copyright 2023 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef COBALT_JS_PROFILER_PROFILER_H_
+#define COBALT_JS_PROFILER_PROFILER_H_
+
+#include <memory>
+#include <string>
+
+#include "cobalt/dom/performance_high_resolution_time.h"
+#include "cobalt/js_profiler/profiler_init_options.h"
+#include "cobalt/js_profiler/profiler_trace.h"
+#include "cobalt/script/promise.h"
+#include "cobalt/script/value_handle.h"
+#include "cobalt/script/wrappable.h"
+#include "cobalt/web/event_target.h"
+#include "third_party/v8/include/cppgc/member.h"
+#include "third_party/v8/include/v8-profiler.h"
+
+namespace cobalt {
+namespace js_profiler {
+
+class Profiler : public cobalt::web::EventTarget {
+ public:
+  using ProfilerTracePromise = script::HandlePromiseWrappable;
+
+  Profiler(script::EnvironmentSettings* settings, ProfilerInitOptions options,
+           script::ExceptionState* exception_state);
+
+  ProfilerTracePromise Stop(script::EnvironmentSettings* environment_settings);
+
+  bool stopped() const { return stopped_; }
+
+  dom::DOMHighResTimeStamp sample_interval() const { return sample_interval_; }
+
+  DEFINE_WRAPPABLE_TYPE(Profiler);
+
+  virtual v8::CpuProfilingStatus ImplProfilingStart(
+      std::string profiler_id, v8::CpuProfilingOptions options,
+      script::EnvironmentSettings* settings);
+
+ private:
+  void PerformStop(script::EnvironmentSettings* environment_settings,
+                   std::unique_ptr<script::ValuePromiseWrappable::Reference>
+                       promise_reference,
+                   base::TimeTicks time_origin, std::string profiler_id);
+
+  std::string nextProfileId();
+
+  bool stopped_;
+  dom::DOMHighResTimeStamp sample_interval_;
+  v8::CpuProfiler* cpu_profiler_ = nullptr;
```

Comment:
I can't use scoped_refptr, because the v8::CpuProfiler class doesn't implement AddRef. I also get this error when I use unique_ptr, saying that I'm calling a private destructor.

```
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/unique_ptr.h:99:9: error: calling a private destructor of class 'v8::CpuProfiler'
        delete __ptr;
               ^
/usr/lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/unique_ptr.h:404:4: note: in instantiation of member function 'std::default_delete<v8::CpuProfiler>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
../../cobalt/js_profiler/profiler.cc:48:11: note: in instantiation of member function 'std::unique_ptr<v8::CpuProfiler>::~unique_ptr' requested here
Profiler::Profiler(script::EnvironmentSettings* settings,
          ^
../../third_party/v8/include/v8-profiler.h:400:3: note: declared private here
  ~CpuProfiler();
  ^
1 error generated.
ninja: build stopped: subcommand failed.
```

Reply:
Unfortunately I can't seem to make a unique_ptr from an object with a private destructor (at least not without a custom Deleter). So I added a destructor to the Profiler class, which should be called by JS Garbage collection when appropriate. Please let me know if this is sufficient, thank you.